<a href="https://colab.research.google.com/github/dutt2022/Coursera-Data-science-Projects/blob/main/xgbboost_9_targests.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:

# IMPORTANT: RUN THIS CELL IN ORDER TO IMPORT YOUR KAGGLE DATA SOURCES
# TO THE CORRECT LOCATION (/kaggle/input) IN YOUR NOTEBOOK,
# THEN FEEL FREE TO DELETE THIS CELL.
# NOTE: THIS NOTEBOOK ENVIRONMENT DIFFERS FROM KAGGLE'S PYTHON
# ENVIRONMENT SO THERE MAY BE MISSING LIBRARIES USED BY YOUR
# NOTEBOOK.

import os
import sys
from tempfile import NamedTemporaryFile
from urllib.request import urlopen
from urllib.parse import unquote, urlparse
from urllib.error import HTTPError
from zipfile import ZipFile
import tarfile
import shutil

CHUNK_SIZE = 40960
DATA_SOURCE_MAPPING = 'financial-performance-prediction:https%3A%2F%2Fstorage.googleapis.com%2Fkaggle-competitions-data%2Fkaggle-v2%2F70781%2F7733789%2Fbundle%2Farchive.zip%3FX-Goog-Algorithm%3DGOOG4-RSA-SHA256%26X-Goog-Credential%3Dgcp-kaggle-com%2540kaggle-161607.iam.gserviceaccount.com%252F20240307%252Fauto%252Fstorage%252Fgoog4_request%26X-Goog-Date%3D20240307T112846Z%26X-Goog-Expires%3D259200%26X-Goog-SignedHeaders%3Dhost%26X-Goog-Signature%3D1276e621ede41c3556cfafa0324160b0fdff1c3f6bfd91e89b9629823a7eff58303299e4dca367e489c4d62930755f952b3be91bd5fa926715d2bdfd2a002df4636861397dd62abbb3c437e500d3b8de7ad6725a674c3168f12f80fd1e6a0d66cf0291bf9864dd6f70035c2a7efe4777d6ea06f54f79f44b166085518e79fb0957d7944433e864b9babd87344d6813320947fb6dfdaf43a197261bf95f48b52906145bab5345739b837b0f0ce679fa1c1c54c5d3b7796db098f75bff16c955e59feb5c59f687659c78ba8ada5a591cc96793e485d4ca699d4c73ddeb5d7f5dd89745c1fa9cf1c591ef59bbb7356e4fc3e930eaeaf585c218cb32118163d2d9a8'

KAGGLE_INPUT_PATH='/kaggle/input'
KAGGLE_WORKING_PATH='/kaggle/working'
KAGGLE_SYMLINK='kaggle'

!umount /kaggle/input/ 2> /dev/null
shutil.rmtree('/kaggle/input', ignore_errors=True)
os.makedirs(KAGGLE_INPUT_PATH, 0o777, exist_ok=True)
os.makedirs(KAGGLE_WORKING_PATH, 0o777, exist_ok=True)

try:
  os.symlink(KAGGLE_INPUT_PATH, os.path.join("..", 'input'), target_is_directory=True)
except FileExistsError:
  pass
try:
  os.symlink(KAGGLE_WORKING_PATH, os.path.join("..", 'working'), target_is_directory=True)
except FileExistsError:
  pass

for data_source_mapping in DATA_SOURCE_MAPPING.split(','):
    directory, download_url_encoded = data_source_mapping.split(':')
    download_url = unquote(download_url_encoded)
    filename = urlparse(download_url).path
    destination_path = os.path.join(KAGGLE_INPUT_PATH, directory)
    try:
        with urlopen(download_url) as fileres, NamedTemporaryFile() as tfile:
            total_length = fileres.headers['content-length']
            print(f'Downloading {directory}, {total_length} bytes compressed')
            dl = 0
            data = fileres.read(CHUNK_SIZE)
            while len(data) > 0:
                dl += len(data)
                tfile.write(data)
                done = int(50 * dl / int(total_length))
                sys.stdout.write(f"\r[{'=' * done}{' ' * (50-done)}] {dl} bytes downloaded")
                sys.stdout.flush()
                data = fileres.read(CHUNK_SIZE)
            if filename.endswith('.zip'):
              with ZipFile(tfile) as zfile:
                zfile.extractall(destination_path)
            else:
              with tarfile.open(tfile.name) as tarfile:
                tarfile.extractall(destination_path)
            print(f'\nDownloaded and uncompressed: {directory}')
    except HTTPError as e:
        print(f'Failed to load (likely expired) {download_url} to path {destination_path}')
        continue
    except OSError as e:
        print(f'Failed to load {download_url} to path {destination_path}')
        continue

print('Data source import complete.')


[==================================================] 1436515 bytes downloaded
Downloaded and uncompressed: financial-performance-prediction
Data source import complete.


In [2]:
import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)
import warnings
warnings.simplefilter(action='ignore', category=FutureWarning)
from sklearn.model_selection import cross_val_score as cvs
from xgboost import XGBRegressor

import os
for dirname, _, filenames in os.walk('/kaggle/input'):
    for filename in filenames:
        print(os.path.join(dirname, filename))

/kaggle/input/financial-performance-prediction/data_dictionary.txt
/kaggle/input/financial-performance-prediction/test.csv
/kaggle/input/financial-performance-prediction/sample_submission.csv
/kaggle/input/financial-performance-prediction/train.csv


In [3]:
train = pd.read_csv('/kaggle/input/financial-performance-prediction/train.csv')
test = pd.read_csv('/kaggle/input/financial-performance-prediction/test.csv')

## Look at data shape, dtypes and unite train & test into a single DF

In [4]:
print(f'Train shape: {train.shape}')
print(f'Test shape: {test.shape}')

Train shape: (1624, 212)
Test shape: (406, 203)


In [5]:
targets_names = [col for col in train if col not in test]
print(f'Target columns: {targets_names}')

Target columns: ['Q0_TOTAL_ASSETS', 'Q0_TOTAL_LIABILITIES', 'Q0_TOTAL_STOCKHOLDERS_EQUITY', 'Q0_GROSS_PROFIT', 'Q0_COST_OF_REVENUES', 'Q0_REVENUES', 'Q0_OPERATING_INCOME', 'Q0_OPERATING_EXPENSES', 'Q0_EBITDA']


In [6]:
train.dtypes.value_counts()

float64    207
object       4
int64        1
dtype: int64

In [7]:
test.dtypes.value_counts()

float64    198
object       4
int64        1
dtype: int64

In [8]:
# make an easy col to split back the train and test sets
train['is_train'] = True
test['is_train'] = False
df = pd.concat([train, test], axis = 0, ignore_index = True)

## Deal with string/categoric columns

In [9]:
df.select_dtypes(include = 'O')

industry              sector  \
0                            Personal Services   Consumer Cyclical   
1                Building Products & Equipment         Industrials   
2                                          NaN             Unknown   
3           Scientific & Technical Instruments          Technology   
4     Drug Manufacturers - Specialty & Generic          Healthcare   
...                                        ...                 ...   
2025                            Packaged Foods  Consumer Defensive   
2026                                       NaN             Unknown   
2027                           Medical Devices          Healthcare   
2028  Drug Manufacturers - Specialty & Generic          Healthcare   
2029                       Specialty Chemicals     Basic Materials   

     recommendationKey financialCurrency  
0           strong_buy               USD  
1                  buy               USD  
2                  NaN               NaN  
3                 hold               USD  
4                  buy               USD  
...                ...               ...  
2025              hold               USD  
2026               NaN               NaN  
2027               buy               USD  
2028               buy               USD  
2029               buy               USD  

[2030 rows x 4 columns]

In [10]:
for col in df.select_dtypes(include = 'O'):
    print(f'{col} unique values num: {df[col].nunique()}')
    print('-'*50)

industry unique values num: 113
--------------------------------------------------
sector unique values num: 10
--------------------------------------------------
recommendationKey unique values num: 6
--------------------------------------------------
financialCurrency unique values num: 2
--------------------------------------------------


Looks like it is best have to label encode the `industry` with 113 unique categories.
`sector` and `recommendationKey` can safely be one-hot-encoded as it will expand the dataset by only 16 columns.
And we will just binarize the `financialCurrency` column

In [11]:
def binarize(df, col):
    uniques = list(df[col].dropna().unique())
    df[col] = [1 if val == uniques[0] else 0 for val in df[col]]
    return df

In [12]:
def one_hot_encode(df, col):
    dummies = pd.get_dummies(df[col], prefix = col).astype(int)
    df.drop(col, axis = 1, inplace = True)
    df = pd.concat([df, dummies], axis = 1)
    return df

In [13]:
def factorize(df, col):
    mapping = {name: ix for ix, name in enumerate(df[col].unique())}
    df[col] = df[col].map(mapping)
    return df

In [14]:
df = binarize(df, 'financialCurrency')
print(df['financialCurrency'].head())

0    1
1    1
2    0
3    1
4    1
Name: financialCurrency, dtype: int64


In [15]:
print(df.shape)
df = one_hot_encode(df, 'sector')
df = one_hot_encode(df, 'recommendationKey')
print(df.shape)

(2030, 213)
(2030, 227)


In [16]:
df = factorize(df, 'industry')
print(df.dtypes.value_counts())

float64    207
int64       19
bool         1
dtype: int64


In [17]:
import warnings
warnings.simplefilter(action='ignore', category=FutureWarning)

## Deal with missing values

In [18]:
df.isnull().sum()

Id                                  0
industry                            0
fullTimeEmployees                 102
auditRisk                         625
boardRisk                         625
                                 ... 
recommendationKey_hold              0
recommendationKey_none              0
recommendationKey_sell              0
recommendationKey_strong_buy        0
recommendationKey_underperform      0
Length: 227, dtype: int64

In [19]:
print('Missing data stats:\n')
under_10_percent = 0
ten_to_20_percent = 0
twenty_to_50_percent = 0
over_50_percent = 0
for col in df:
    if np.any(df[col].isnull()):
        percent_missing = np.round(100*df[col].isnull().sum()/len(df),2)
        if percent_missing < 10:
            under_10_percent+=1
        elif 10 <= percent_missing < 20:
            ten_to_20_percent+=1
        elif 20 <= percent_missing < 50:
            twenty_to_50_percent+=1
        else:
            over_50_percent+=1

print(f'ncols with under 10% missing data:\t{under_10_percent}')
print(f'ncols with 10-20% missing data:\t\t{ten_to_20_percent}')
print(f'ncols with 20-50% missing data:\t\t{twenty_to_50_percent}')
print(f'ncols with over 50% missing data:\t{over_50_percent}')

Missing data stats:

ncols with under 10% missing data:	134
ncols with 10-20% missing data:		44
ncols with 20-50% missing data:		16
ncols with over 50% missing data:	1


Many columns have missing data, but the majority of cols with NaNs have under 10% missing values. It's safe to fill all missing values by median.

In [20]:
# check if there are any inf values in data
df[df == np.inf].count().sum()

17

In [21]:
# check for missing values in the target columns
train[targets_names].isnull().sum()

Q0_TOTAL_ASSETS                 0
Q0_TOTAL_LIABILITIES            0
Q0_TOTAL_STOCKHOLDERS_EQUITY    0
Q0_GROSS_PROFIT                 0
Q0_COST_OF_REVENUES             0
Q0_REVENUES                     0
Q0_OPERATING_INCOME             0
Q0_OPERATING_EXPENSES           0
Q0_EBITDA                       0
dtype: int64

In [22]:
def impute_missing_by_median(df):
    print(f'Missing values before imputation: {sum(df.isnull().sum())}')
    # first replace inf and -inf with nan
    df.replace([np.inf, -np.inf], np.nan, inplace=True)
    # then impute nan by median
    for col in df:
        if np.any(df[col].isnull()):
            df[col].fillna(df[col].median(), inplace = True)
    print(f'Missing values after imputation: {sum(df.isnull().sum())}')
    return df

In [23]:
df = impute_missing_by_median(df)

Missing values before imputation: 28653
Missing values after imputation: 0


## Extract train & test without target cols and separate tartgets df with 9 target cols

In [24]:
train = df[df['is_train'] == True]
test = df[df['is_train'] == False]
train.drop('is_train', axis = 1, inplace=True)
test.drop('is_train', axis = 1, inplace=True)
test.reset_index(drop = True, inplace = True)

test.drop(targets_names, axis = 1, inplace = True)
targets = train[targets_names].copy()
train.drop(targets_names, axis = 1, inplace = True)

<ipython-input-24-996949eefa5f>:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  train.drop('is_train', axis = 1, inplace=True)
<ipython-input-24-996949eefa5f>:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  test.drop('is_train', axis = 1, inplace=True)
<ipython-input-24-996949eefa5f>:7: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  test.drop(targets_names, axis = 1, inplace = True)
<ipython-input-24-996949eefa5f>:9: SettingWithCopyWa

## Validate XGB model on train data
### Since we have 9 independent target columns, I will train 9 separate separate models on train features to predict each target.

In [25]:
targets

Q0_TOTAL_ASSETS  Q0_TOTAL_LIABILITIES  Q0_TOTAL_STOCKHOLDERS_EQUITY  \
0        1.192950e+09          1.055816e+09                  1.371340e+08   
1        1.219954e+09          8.350460e+08                  3.849080e+08   
2       -1.358200e+11         -3.797200e+10                 -9.784800e+10   
3        8.329800e+07          1.604000e+07                 -3.163300e+07   
4        1.485460e+08          5.655500e+07                  9.199100e+07   
...               ...                   ...                           ...   
1619     3.539600e+08          1.997590e+08                  1.542010e+08   
1620     1.061530e+08          3.949300e+07                  6.811700e+07   
1621     5.203150e+08          3.807960e+08                  3.233080e+08   
1622     1.997210e+08          2.692100e+07                  1.728000e+08   
1623     1.333626e+10          1.504531e+10                 -8.217640e+08   

      Q0_GROSS_PROFIT  Q0_COST_OF_REVENUES   Q0_REVENUES  Q0_OPERATING_INCOME  \
0          25598000.0         6.189900e+07  4.299200e+07         1.522200e+07   
1         111786000.0         2.177960e+08  3.295820e+08         2.890800e+07   
2        -435000000.0         4.445000e+09  4.010000e+09        -1.472000e+09   
3            494000.0         3.339000e+06  3.833000e+06        -1.234800e+07   
4           1761000.0        -2.375000e+06 -6.140000e+05        -1.009800e+07   
...               ...                  ...           ...                  ...   
1619       48741000.0         4.569300e+07  9.443400e+07         2.616000e+06   
1620        1077000.0         1.690000e+05  1.246000e+06        -6.705000e+06   
1621       27808000.0         1.123840e+08  1.401920e+08         7.061000e+06   
1622              0.0         0.000000e+00  0.000000e+00        -3.391900e+07   
1623      394720000.0         5.777330e+08  9.724530e+08         6.259500e+07   

      Q0_OPERATING_EXPENSES    Q0_EBITDA  
0                10383000.0   15215000.0  
1                85522000.0   26264000.0  
2               219000000.0 -654000000.0  
3                12842000.0  -12348000.0  
4                11859000.0  -10098000.0  
...                     ...          ...  
1619             46125000.0    2616000.0  
1620              7782000.0   -6705000.0  
1621              9417000.0   18391000.0  
1622             33919000.0  -33919000.0  
1623            268445000.0  126275000.0  

[1624 rows x 9 columns]

In [26]:
model = XGBRegressor()

In [27]:
print('Cross validation R2 scores for each target:\n')
cross_val_score_results = {}
for target in targets:
    score = np.round(np.mean(cvs(model, train, targets[target], cv=10, scoring='r2')),2)
    cross_val_score_results[target] = score
    print(f'{target} -> {score}')
print(f'\nMean R2 score across all targets: {np.mean(list(cross_val_score_results.values()))}')

Cross validation R2 scores for each target:

Q0_TOTAL_ASSETS -> 0.83
Q0_TOTAL_LIABILITIES -> 0.58
Q0_TOTAL_STOCKHOLDERS_EQUITY -> 0.9
Q0_GROSS_PROFIT -> 0.89
Q0_COST_OF_REVENUES -> 0.9
Q0_REVENUES -> 0.95
Q0_OPERATING_INCOME -> 0.78
Q0_OPERATING_EXPENSES -> 0.88
Q0_EBITDA -> 0.01

Mean R2 score across all targets: 0.7466666666666667


Arbitrary results. EBITDA R2 is negative, but the rest are not bad. Let's try to upload a first submission.

## Train same model on the whole trains set, predict on test, save to dictionary

In [28]:
preds = {}
for target in targets:
    model.fit(train, targets[target])
    pred = model.predict(test)
    preds[target] = pred
    print(f'Finished train/predict for: {target}')

Finished train/predict for: Q0_TOTAL_ASSETS
Finished train/predict for: Q0_TOTAL_LIABILITIES
Finished train/predict for: Q0_TOTAL_STOCKHOLDERS_EQUITY
Finished train/predict for: Q0_GROSS_PROFIT
Finished train/predict for: Q0_COST_OF_REVENUES
Finished train/predict for: Q0_REVENUES
Finished train/predict for: Q0_OPERATING_INCOME
Finished train/predict for: Q0_OPERATING_EXPENSES
Finished train/predict for: Q0_EBITDA


## Make submission

In [29]:
sub = pd.read_csv('/kaggle/input/financial-performance-prediction/sample_submission.csv')

In [30]:
# sanity check
sub.Id == test.Id

0      True
1      True
2      True
3      True
4      True
       ... 
401    True
402    True
403    True
404    True
405    True
Name: Id, Length: 406, dtype: bool

In [31]:
for target in preds:
    sub[target] = preds[target]
sub.head()

Id   Q0_REVENUES  Q0_COST_OF_REVENUES  Q0_GROSS_PROFIT  \
0   480 -1.211120e+06        -3.895045e+05    -5.017264e+04   
1   573  5.949692e+08         3.356995e+08     1.815985e+08   
2  1967 -1.272589e+06         1.030028e+07    -6.526061e+05   
3    57  4.606910e+08         2.904094e+08     1.891734e+08   
4   131  9.713140e+08         8.411632e+08     1.933180e+08   

   Q0_OPERATING_EXPENSES     Q0_EBITDA  Q0_OPERATING_INCOME  Q0_TOTAL_ASSETS  \
0          -7.105477e+05 -1.565854e+06        -1.995759e+06    -2.966194e+07   
1           1.065167e+08  1.317824e+08         4.421544e+07     3.533268e+09   
2           4.120479e+06 -6.113872e+06        -5.567920e+06     5.888865e+07   
3           7.999364e+07  6.343608e+07         6.295966e+07     6.022530e+09   
4           8.722598e+07  1.275100e+08         1.266720e+08     4.397583e+09   

   Q0_TOTAL_LIABILITIES  Q0_TOTAL_STOCKHOLDERS_EQUITY  
0         -1.779459e+06                  4.574116e+05  
1          1.871350e+09                  5.346400e+08  
2          2.424615e+07                  1.971546e+06  
3          2.713045e+09                  3.356197e+09  
4          2.879660e+09                  1.119394e+09

In [32]:
sub.to_csv('submission.csv', index = False)

In [33]:
#ids = np.array(sub.Id)
#preds.keys()

In [34]:
# Write the submission file
#np.savetxt(
#    'submission.csv',
#    np.rec.fromarrays([ids] + [pred for pred in list(preds.values())]),
#    fmt=['%s', '%d', '%d', '%d', '%d', '%d', '%d', '%d', '%d', '%d'],
#    delimiter=',',
#    header='Id,Q0_TOTAL_ASSETS,Q0_TOTAL_LIABILITIES,Q0_TOTAL_STOCKHOLDERS_EQUITY,Q0_GROSS_PROFIT,Q0_COST_OF_REVENUES,Q0_REVENUES,Q0_OPERATING_INCOME,Q0_OPERATING_EXPENSES,Q0_EBITDA',
#    comments='submission description from code',
#)